### PR TITLE
Simplify API hierarchy of HTTP and gRPC client builders

### PR DESCRIPTION
**Pull request by chemicL**
Motivation:

The current hierarchy of HTTP and gRPC client builders is complicated
and lacks alignment between both stacks. The HTTP hierarchy in
particular carries along some historical context which was recently
updated to make the `Partitioned`- and `MultiAddress`- `HttpClientBuilder`s
independent of the API useful for `SingleAddressHttpClientBuilder` as
keeping them consistent was problematic and led to the introduction of
`SingleAddressInitializer` mechanism in both cases. After recent
untangling of particular clients' dependency on those shared APIs it's
now time to simplify the hierarchy, which this change attempts.

Modifications:

- Removed `BaseGrpcClientBuilder`,
- Removed `SingleAddressGrpcClientBuilder`,
- Removed `BaseSingleAddressHttpClientBuilder`,
- Moved methods of `HttpClientBuilder` into
  `SingleAddressHttpClientBuilder` to highlight their meaning in the
  context of a client that is dedicated to a particular address. This
  change left the `HttpClientBuilder` interface empty,
- Merged `ExecutionContextAwareHttpBuilder` and
  `HttpClientBuildFinalizer` into `HttpClientBuilder` as those
  package-private interfaces were useful in the untangling process. Now
  they're not required any more and can be incorporated into
  `HttpClientBuilder` which solely represents the responsibilities that
  those interfaces encapsulated,
- Reverted `MultiAddressHttpClientBuilder` and
  `PartitionedHttpClientBuilder` to abstract classes instead of
  interfaces to remain consistent with the server builders and the gRPC
  counterparts.

Result:

The API hierarchy is simplified, uses similar concepts for different
levels across the HTTP and gRPC stack and the more complex multi- and
partitioned- client builders have a common base interface that doesn't
carry unnecessary mechanisms from the single address use case. At the
same time, for most usages where the `HttpClients` and `GrpcClients`
factories are used, there would be no action required from the users.

### CODE REVIEW

**Review by bondolo — COMMENTED** [review]
Great simplification. I don't feel confident to approve though

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java [hunk 1]
```diff
@@ -50,72 +58,235 @@
...
+     * Append the filter to the chain of filters used to decorate the {@link ConnectionFactory} used by this
+     * builder.
+     * <p>
+     * Filtering allows you to wrap a {@link ConnectionFactory} and modify behavior of
+     * {@link ConnectionFactory#newConnection(Object, TransportObserver)}.
+     * Some potential candidates for filtering include logging and metrics.
+     * <p>
+     * The order of execution of these filters are in order of append. If 3 filters are added as follows:
+     * <pre>
+     *     builder.append(filter1).append(filter2).append(filter3)
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:143]: To be consistent with HTTP, please use the real method name here: `appendConnectionFactoryFilter`

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java [hunk 2]
```diff
@@ -50,72 +58,235 @@
...
+     * Append the filter to the chain of filters used to decorate the {@link StreamingHttpConnection} created by this
+     * builder.
+     * <p>
+     * Filtering allows you to wrap a {@link StreamingHttpConnection} and modify behavior during request/response
+     * processing
+     * Some potential candidates for filtering include logging, metrics, and decorating responses.
+     * <p>
+     * The order of execution of these filters are in order of append. If 3 filters are added as follows:
+     * <pre>
+     *     builder.append(filter1).append(filter2).append(filter3)
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:165]: To be consistent with HTTP, please use the real method name here: `appendConnectionFilter`

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java [hunk 3]
```diff
@@ -50,72 +58,235 @@
...
+     * Append the filter to the chain of filters used to decorate the {@link StreamingHttpConnection} created by this
+     * builder, for every request that passes the provided {@link Predicate}.
+     * <p>
+     * Filtering allows you to wrap a {@link StreamingHttpConnection} and modify behavior during request/response
+     * processing
+     * Some potential candidates for filtering include logging, metrics, and decorating responses.
+     * <p>
+     * The order of execution of these filters are in order of append. If 3 filters are added as follows:
+     * <pre>
+     *     builder.append(filter1).append(filter2).append(filter3)
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:188]: To be consistent with HTTP, please use the real method name here: `appendConnectionFilter`.

Same for other `append*` methods we have below (pre-existing, I can not comment there).

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java [hunk 4]
```diff
@@ -50,72 +58,235 @@
...
     public abstract GrpcClientBuilder<U, R> appendConnectionFilter(Predicate<StreamingHttpRequest> predicate,
                                                                    StreamingHttpConnectionFilterFactory factory);
 
-    @Override
+    /**
+     * Set the SSL/TLS configuration.
+     * @param sslConfig The configuration to use.
+     * @return {@code this}.
+     */
     public abstract GrpcClientBuilder<U, R> sslConfig(ClientSslConfig sslConfig);
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:210]: Consider adding `@see ClientSslConfigBuilder` tag

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java [hunk 5]
```diff
@@ -50,72 +58,235 @@
...
+     * the clients automatically when allowed by the passed {@link AutoRetryStrategyProvider}. These retries are not a
+     * substitute for user level retries which are designed to infer retry decisions based on request/error information.
+     * Typically such user level retries are done using filters but can also be done differently per request
+     * (eg: by using {@link Single#retry(BiIntPredicate)}).
+     *
+     * @param autoRetryStrategyProvider {@link AutoRetryStrategyProvider} for the automatic retry strategy.
+     * @return {@code this}
+     */
     public abstract GrpcClientBuilder<U, R> autoRetryStrategy(
             AutoRetryStrategyProvider autoRetryStrategyProvider);
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:248]: Consider adding `@see DefaultAutoRetryStrategyProvider` tag

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java [hunk 6]
```diff
@@ -50,72 +58,235 @@
...
     public abstract GrpcClientBuilder<U, R> bufferAllocator(BufferAllocator allocator);
 
-    @Override
+    /**
+     * Sets the {@link GrpcExecutionStrategy} for all clients created from this builder.
+     *
+     * @param strategy {@link GrpcExecutionStrategy} to use.
+     * @return {@code this}.
+     */
     public abstract GrpcClientBuilder<U, R> executionStrategy(GrpcExecutionStrategy strategy);
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:87]: Consider adding `@see GrpcExecutionStrategies` tag

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java [hunk 7]
```diff
@@ -50,72 +58,235 @@
...
+    /**
+     * Configurations of various underlying protocol versions.
+     * <p>
+     * <b>Note:</b> the order of specified protocols will reflect on priorities for ALPN in case the connections are
+     * secured.
+     *
+     * @param protocols {@link HttpProtocolConfig} for each protocol that should be supported.
+     * @return {@code this}.
+     */
     public abstract GrpcClientBuilder<U, R> protocols(HttpProtocolConfig... protocols);
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:122]: Consider adding `@see HttpProtocolConfigs` tag

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java [hunk 8]
```diff
@@ -50,72 +58,235 @@
...
     public abstract GrpcClientBuilder<U, R> serviceDiscoverer(
             ServiceDiscoverer<U, R, ServiceDiscovererEvent<R>> serviceDiscoverer);
 
-    @Override
+    /**
+     * Set a {@link HttpLoadBalancerFactory} to create {@link LoadBalancer} instances.
+     *
+     * @param loadBalancerFactory {@link HttpLoadBalancerFactory} to create {@link LoadBalancer} instances.
+     * @return {@code this}.
+     */
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:289]: Can not comment on line 381:
Interesting that we do not have any test or example for `buildMulti()`. We should revisit if it's useful or not.

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpClientBuilder.java [hunk 9]
```diff
@@ -45,256 +26,63 @@
...
-     * Calling {@link ConnectionFactory} wrapped by this filter chain, the order of invocation of these filters will be:
-     * <pre>
-     *     filter1 ⇒ filter2 ⇒ filter3 ⇒ original connection factory
-     * </pre>
-     * @param factory {@link ConnectionFactoryFilter} to use.
-     * @return {@code this}
-     */
-    public abstract HttpClientBuilder<U, R, SDE> appendConnectionFactoryFilter(
-            ConnectionFactoryFilter<R, FilterableStreamingHttpConnection> factory);
+    HttpClientBuilder<U, R, SDE> executionStrategy(HttpExecutionStrategy strategy);
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpClientBuilder.java:53]: Consider adding `@see HttpExecutionStrategies` tag

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java [hunk 10]
```diff
@@ -39,88 +49,255 @@
...
+    /**
+     * Configurations of various HTTP protocol versions.
+     * <p>
+     * <b>Note:</b> the order of specified protocols will reflect on priorities for
+     * <a href="https://tools.ietf.org/html/rfc7301">ALPN</a> in case the connections use TLS.
+     *
+     * @param protocols {@link HttpProtocolConfig} for each protocol that should be supported.
+     * @return {@code this}.
+     */
     public abstract SingleAddressHttpClientBuilder<U, R> protocols(HttpProtocolConfig... protocols);
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:88]: Consider adding `@see HttpProtocolConfigs` tag

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java [hunk 11]
```diff
@@ -39,88 +49,255 @@
...
+     * substitute for user level retries which are designed to infer retry decisions based on request/error information.
+     * Typically such user level retries are done using filters (eg:
+     * {@link #appendClientFilter(StreamingHttpClientFilterFactory)}) but can also be done differently per request
+     * (eg: by using {@link Single#retry(BiIntPredicate)}).
+     *
+     * @param autoRetryStrategyProvider {@link AutoRetryStrategyProvider} for the automatic retry strategy.
+     * @return {@code this}
+     */
+    public abstract SingleAddressHttpClientBuilder<U, R> autoRetryStrategy(
+            AutoRetryStrategyProvider autoRetryStrategyProvider);
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:269]: Consider adding `@see DefaultAutoRetryStrategyProvider` tag

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java [hunk 12]
```diff
@@ -39,88 +49,255 @@
...
+            ServiceDiscoveryRetryStrategy<R, ServiceDiscovererEvent<R>> retryStrategy);
+
+    /**
+     * Sets a {@link HttpLoadBalancerFactory} to create {@link LoadBalancer} instances.
+     *
+     * @param loadBalancerFactory {@link HttpLoadBalancerFactory} to create {@link LoadBalancer} instances.
+     * @return {@code this}.
+     */
+    public abstract SingleAddressHttpClientBuilder<U, R> loadBalancerFactory(
+            HttpLoadBalancerFactory<R> loadBalancerFactory);
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:300]: Comment for the line 307:
Consider adding `@see ClientSslConfigBuilder` tag

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultPartitionedHttpClientBuilder.java [hunk 13]
```diff
@@ -62,7 +62,7 @@
 import static java.util.Objects.requireNonNull;
 import static java.util.function.Function.identity;
 
-class DefaultPartitionedHttpClientBuilder<U, R> implements PartitionedHttpClientBuilder<U, R> {
+class DefaultPartitionedHttpClientBuilder<U, R> extends PartitionedHttpClientBuilder<U, R> {
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultPartitionedHttpClientBuilder.java:65]: This class can be `final`

**Review by idelpivnitskiy — APPROVED** [review]
Great simplification!
Minor suggestions, otherwise LGTM

**Review by tkountis — APPROVED** [review]
Nice work!!

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java (hunk 7 above)
Comment (chemicL) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:122]: For that to be visible, the grpc-api module would need to depend on http-netty. Do we want that?

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java (hunk 10 above)
Comment (chemicL) [reply servicetalk-http-api/src/main/java/io/servicetalk/http/api/SingleAddressHttpClientBuilder.java:88]: That would require the api module to depend on http-netty.

**Comment by chemicL** [comment]
Thanks @bondolo @idelpivnitskiy @tkountis for the review. I addressed most comments, adding changes where there's no explanation why it wouldn't work.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java (hunk 7 above)
Comment (idelpivnitskiy) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcClientBuilder.java:123]: No, that was overlook, I thought it's in api

**Pull request merged**